File: test-solutions/online-mst-decisions-frontier-cs-algorithmic-153/solution.cpp
#include <iostream>
#include <numeric>
#include <utility>
#include <vector>

using namespace std;

namespace {

constexpr int N = 400;
constexpr int M = 5 * (N - 1);

struct DSU {
    vector<int> parent;
    vector<int> size;

    explicit DSU(int n) : parent(n), size(n, 1) {
        iota(parent.begin(), parent.end(), 0);
    }

    int find(int value) {
        while (parent[value] != value) {
            parent[value] = parent[parent[value]];
            value = parent[value];
        }
        return value;
    }

    bool unite(int left, int right) {
        left = find(left);
        right = find(right);
        if (left == right) return false;
        if (size[left] < size[right]) swap(left, right);
        parent[right] = left;
        size[left] += size[right];
        return true;
    }
};

bool solve_case() {
    int x = 0;
    int y = 0;
    for (int i = 0; i < N; ++i) {
        if (!(cin >> x >> y)) return false;
    }

    vector<pair<int, int>> edges(M);
    for (int i = 0; i < M; ++i) {
        if (!(cin >> edges[i].first >> edges[i].second)) return false;
    }

    DSU accepted(N);
    for (const auto& [u, v] : edges) {
        long long cost = 0;
        if (!(cin >> cost)) return false;
        cout << (accepted.unite(u, v) ? 1 : 0) << '\n' << flush;
    }

    return true;
}

}  // namespace

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    while (solve_case()) {
    }

    return 0;
}

File: test-solutions/online-mst-decisions-frontier-cs-algorithmic-153/solution.py
from __future__ import annotations

import sys

N = 400
M = 5 * (N - 1)


class DSU:
    def __init__(self, n: int) -> None:
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, value: int) -> int:
        while self.parent[value] != value:
            self.parent[value] = self.parent[self.parent[value]]
            value = self.parent[value]
        return value

    def union(self, left: int, right: int) -> bool:
        left_root = self.find(left)
        right_root = self.find(right)
        if left_root == right_root:
            return False
        if self.size[left_root] < self.size[right_root]:
            left_root, right_root = right_root, left_root
        self.parent[right_root] = left_root
        self.size[left_root] += self.size[right_root]
        return True


def read_pair() -> tuple[int, int] | None:
    line = sys.stdin.readline()
    if not line:
        return None
    a, b = map(int, line.split())
    return a, b


def main() -> int:
    for _ in range(N):
        if read_pair() is None:
            return 0

    edges: list[tuple[int, int]] = []
    for _ in range(M):
        pair = read_pair()
        if pair is None:
            return 0
        edges.append(pair)

    dsu = DSU(N)
    for u, v in edges:
        cost_line = sys.stdin.readline()
        if not cost_line:
            return 0
        take = 1 if dsu.union(u, v) else 0
        print(take, flush=True)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
